### PR TITLE
[YUNIKORN-2366] Update GetPodResources() to handle in-place pod resource updates

### PR DESCRIPTION
**Pull request by craigcondit**
### What is this PR for?
Implement initial support for the upcoming Kubernetes InPlacePodVerticalScaling feature when doing Pod resource calculations. If present, the allocated resources of each container will be used in preference to the requested resources. Additionally, if the Pod status flag Resize is set to Proposed, use the larger of both requested and allocated for each resource type.

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [x] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2366

### How should this be tested?
Added unit tests to cover logic introduced by the upstream K8s feature.

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/common/resource.go [hunk 1]
```diff
@@ -87,6 +86,48 @@ func GetPodResource(pod *v1.Pod) (resource *si.Resource) {
...
+	// alpha: v1.27
+	// beta: v1.31?
+	// If AllocatedResources are present, these need to be used in preference to pod resource requests.
+	// Additionally, if the Resize pod status is InProgress, then the maximum of the request and allocated values need
+	// to be used.
+	requested := pod.Spec.Containers[i].Resources.Requests
+	if len(pod.Status.ContainerStatuses) == 0 {
+		return getResource(requested)
+	}
+	allocated := pod.Status.ContainerStatuses[i].AllocatedResources
```
Comment (chia7712) [inline pkg/common/resource.go:100]: Do `ContainerStatuses` and `Containers`  have same order?

Comment (craigcondit) [reply pkg/common/resource.go:100]: Yes, this is part of the spec.

Comment (craigcondit) [reply pkg/common/resource.go:100]: See https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#scheduler-and-api-server-interaction for more details.

File: pkg/common/resource.go [hunk 2]
```diff
@@ -87,6 +86,48 @@ func GetPodResource(pod *v1.Pod) (resource *si.Resource) {
...
+	if len(allocated) == 0 {
+		// no allocatedResources present, use requested
+		return getResource(requested)
+	}
+	if pod.Status.Resize == v1.PodResizeStatusProposed {
```
Comment (chia7712) [inline pkg/common/resource.go:105]: from the spec: `For containers which have Status.Resize = "Proposed", it must be pessimistic and assume that the resize will be imminently accepted. Therefore it must use the larger of the Pod's Spec...Resources.Requests and Status...AllocatedResources values`

Should we replace `PodResizeStatusInProgress` by `PodResizeStatusProposed`?

Comment (chia7712) [reply pkg/common/resource.go:105]: oh, you have updated the code. please ignore my comment.

Comment (craigcondit) [reply pkg/common/resource.go:105]: Yes, I spotted that while re-reading the spec.

**Review by chia7712 — APPROVED** [review]
LGTM

**Review by pbacsko — APPROVED** [review]
+1

**Pull request closed without merging**